File: Day37/Part1/minimum_window_substring.py
from collections import Counter

def min_window(s: str, t: str) -> str:
    if not t or not s:
        return ""
    need = Counter(t)
    required = len(need)
    formed = 0
    window = {}
    l = 0
    ans = (float('inf'), None, None)
    for r, ch in enumerate(s):
        window[ch] = window.get(ch, 0) + 1
        if ch in need and window[ch] == need[ch]:
            formed += 1
        while l <= r and formed == required:
            if r - l + 1 < ans[0]:
                ans = (r - l + 1, l, r)
            lc = s[l]
            window[lc] -= 1
            if lc in need and window[lc] < need[lc]:
                formed -= 1
            l += 1
    return "" if ans[0] == float('inf') else s[ans[1]:ans[2]+1]

if __name__ == "__main__":
    print(min_window("ADOBECODEBANC", "ABC"))  # Expected "BANC"

File: learnLeetCode/Day37/Part1/minimum_window_substring.cpp
#include <bits/stdc++.h>
using namespace std;

string minWindow(string s, string t){
    if(t.empty()||s.empty()) return "";
    vector<int> need(128,0), have(128,0);
    int required=0;
    for(char c: t){ if(need[c]==0) ++required; ++need[c]; }
    int formed=0; int l=0; int bestLen=INT_MAX, bestL=0, bestR=-1;
    for(int r=0;r<(int)s.size();++r){
        char c = s[r];
        have[c]++;
        if(need[c]>0 && have[c]==need[c]) formed++;
        while(l<=r && formed==required){
            if(r-l+1<bestLen){ bestLen=r-l+1; bestL=l; bestR=r; }
            char lc=s[l];
            have[lc]--;
            if(need[lc]>0 && have[lc]<need[lc]) formed--;
            l++;
        }
    }
    return bestLen==INT_MAX? string(): s.substr(bestL, bestLen);
}

int main(){ ios::sync_with_stdio(false); cin.tie(nullptr);
    cout << minWindow("ADOBECODEBANC","ABC") << "\n"; // Expected BANC
    return 0; }
